### ISSUE TITLE
Support Relative Starting Addresses

### ISSUE THREAD

**Issue by Mikaz-fr** [facelessuser/HexViewer#32]
First of all the plugin is really promising, I which I could use it too.
But most of the time I'm working with Intel Hex files and there is no compatibility yet.

Intel Hex is a very common hex format (http://en.wikipedia.org/wiki/Intel_HEX) allowing to specify the addresses of data (as opposed to plain binary).

Would it be possible to add support for this hex format?

The following example can be saved as a .hex.
I attached screenshot of the current rendering and the expected one (from another tool).

```
:0200000400CE2C
:100A000000000000000000000000000000000000E6
:100A100000000000000000000000000000000000D6
:100A200000000000000000000000000000000000C6
:100A300000000000000000000000000000000000B6
:100A400000000000000000000000000000000000A6
:100A50000000000000000000000000000000000096
:100A60000000000000000000000000000000000086
:10FFE0000000000000000000000000000000000011
:10FFF0000000000000000000000000000000000001
:0200000400CF2B
:1000000000000000000000000000000000000000F0
:1000100000000000000000000000000000000000E0
:1000200000000000000000000000000000000000D0
:00000001FF
```

![actual](https://cloud.githubusercontent.com/assets/2411679/6715450/cae2474c-cd5b-11e4-83eb-22ebd2ac7925.png)
![expected](https://cloud.githubusercontent.com/assets/2411679/6715449/caba5d4a-cd5b-11e4-811e-b1284e6cb32f.JPG)

**Comment by facelessuser** [comment]
I guess I need to make things clear about what HexViewer was designed to do. It doesn't read hex formats (or that wasn't the intention). It reads a binary file and converts it into a custom hex format I utilize in Sublime.  It is meant to generically parse any binary file.

I output the hex format in a way that I can easily use to scope the text with a custom sublime language file.  I then use those scopes to tell me when I am in an address or hex data or ascii representation.   The `:` to the left and right of the actual data is intentional as the Sublime language file uses regex to parse the view buffer, and the `:` characters were an easy way to tell where the data starts and ends regardless of how it is aligned or how many bytes or spaces are between the address and ascii.

So, yeah, I am using a hex format, but the hex format I use is a custom format that is easy for me to parse.

I never intended HexViewer to support very specific hex representations of binary files, but to convert the binary file and view in a generic format.  Sure it will accept a file with the `.hex` extension if formatted in this custom way.  For that reason, I also wrote myself a script to convert a binary files directly to this format: https://gist.github.com/facelessuser/8967138.

Everything is very generic with HexViewer, to add support for your very specific use case, I would need to generate a binary from your intel format and then parse that.  Sure I **could** do that, but why don't you just use the binary directly with HexViewer?

**Comment by Mikaz-fr** [comment]
Thanks a lot for the precise answer, I know understand better what the plugin is doing.

Converting from Intel-Hex to binary is not an option because doing that would remove the address information.
So maybe a solution would be for me to right a special converter from Intel-Hex to your specific hex format. 
If I manage to find time to do that would there be a clean and easy way to register such a "plugin" to HexViewer? We basically need a way to detect the file type, convert it for open and convert it back on close.
The best case would be to have it as a supplementary package in the package manager that would hook somehow to HexViewer.

Does that sound a good idea to you?

**Comment by facelessuser** [comment]
> Does that sound a good idea to you?

Some work on HexViewer would have to be done first.  Since HexViewer is designed to take bin file and read it, the internal code expects the start of a hex view to start with address 0x00000000.

I had never coded it in a way to handle relative addresses.  Things like the command to jump to an address would no longer work on your files.  The address info status bar would also be incorrect.  The assumption that hexviewer can dump the view to a bin file and have a legitimate bin file would also now be false, but I guess I could let that slide and just say the user should know better than to export such things.  Not sure yet what would happen with a data block terminating early and then followed by another block.

So, I think if this route were to be taken, HexViewer would need to be prepped with some changes first.  Maybe you would separate non-contiguous data blocks into separate hex views before sending them to HexViewer, but HexViewer would still need to be able to handle relative addresses as it does not currently.

**Comment by facelessuser** [comment]
From a HexViewer perspective, I am going to look at this request more as a request for relative address support.  From there, you can build whatever you want around it.  It will still require sequential address, so if you have different blocks of data with gaps in addresses, you will have to split them up into different views.  I will change the title of  this request.

**Comment by Mikaz-fr** [comment]
That would be great yeah.
Any chance to get multiple blocks in the same view? They could be separated by something. But if we have a file with 10 blocks that would mean opening 10 new views...

**Comment by facelessuser** [comment]
> Any chance to get multiple blocks in the same view? They could be separated by something. But if we have a file with 10 blocks that would mean opening 10 new views...

Not in the immediate future, but maybe not ever as well.

HexViewer can read binaries and export them.  It is designed to handle a whole binary.  I would have to add exceptions everywhere to change behaviors across the board.  It isn't really designed to handle binary fragments which is essentially what you want, but it should be easy enough to coerce it into to handling a single fragment. It will treat that fragment as if it were a whole binary file, I just need to pass around an offset for jumping to specific addresses and such.  Multiple fragments will get a bit trickier as I would need to handle not only multiple address range offsets, but gaps in data between blocks and such.  It just isn't something I am interested in doing at this time.

**Comment by Mikaz-fr** [comment]
Understood, we do not have the same use of the plugin and that's fine.
Thanks for the support, I will try to hack around when I get a chance.

**Comment by facelessuser** [comment]
Relative address support is in commit https://github.com/facelessuser/HexViewer/commit/537f8aa3c9950489ae1869df96b742a084885ca2.

It will be in the next release.  I think I got all the address cases, but let me know if I missed anything.  I will consider this issue closed.  If you find any problems, just open a new issue.

Basically, when running a `hex_viewer` command, you can now feed in a `starting_address` as an argument.

**Comment by facelessuser** [comment]
I just realized I need to analyze hex formats that weren't run through `hex_viewer` and determine the starting address from those...bah, I will probably have that later today.

**Comment by facelessuser** [comment]
Okay, now it is complete with commit https://github.com/facelessuser/HexViewer/commit/4ad20b1a252b117ab2c81a2e4140e0a68be1373e.
